subsys/fs/nvs/nvs.c

```diff
@@ -574,17 +587,42 @@ static int nvs_startup(struct nvs_fs *fs)
 		}
 	}
 
+	u32_t data_wra_offs = 0;
 	/* addr contains address of the last ate in the most recent sector
 	 * search for the first ate containing all 0xff
 	 */
 	while (1) {
+		if ((addr & ADDR_OFFS_MASK) < ate_size) {
+			/* return -EDEADLK; */
+			/* sector must contains only delete ate's */
+			break;
+		}
 		addr -= ate_size;
-		rc = nvs_flash_cmp_const(fs, addr, 0xff,
-					  sizeof(struct nvs_ate));
+
```

Comment:
Why all this complicated stuff, just check if the address reached a sector start. If this is the case return an error.

Reply:
I would also say that it's better to fail hard when detecting any inconsistency in the FS structure (without any auto-recovery attempts).

Reply:
There is no auto recovery here. the latest ATE might be just after the latest DATA. This if statement will detect that and will break the while much earlier than `if` in line 595 (than ate scanning come to sector begin). Line 595 is the last resore - it serves damaged FS/not-NVS case, while this one serves just filled up sector.